File: leetcode/p00053/p00053.py
from typing import List

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(1)
    def maxSubArray(self, nums: List[int]) -> int:
        res = -10001
        cur = 0

        for num in nums:
            if cur < 0:
                cur = 0
            
            cur += num
            if cur > res:
                res = cur
        
        return res

if __name__ == '__main__':
    s = Solution()

    assert s.maxSubArray([-2,1,-3,4,-1,2,1,-5,4]) == 6
    assert s.maxSubArray([1]) == 1
    assert s.maxSubArray([5,4,-1,7,8]) == 23

File: leetcode/p00053/p00053.go
package p00053

// Time Complexity: O(n)
// Space Complexity: O(1)
func maxSubArray(nums []int) int {
	// 預設選擇第一個 element 做為初始值
	result := nums[0]

	dp := make([]int, len(nums)+1)
	for i := 1; i <= len(nums); i++ {
		// 每次都可以選擇要繼續累加前面的, 或者是歸零重來
		dp[i] = max(dp[i-1], 0) + nums[i-1]
		result = max(result, dp[i])
	}

	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
